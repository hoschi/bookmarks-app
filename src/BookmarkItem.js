import * as R from 'ramda'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import IconButton from '@material-ui/core/IconButton'
import ListItem from '@material-ui/core/ListItem'
import ListItemText from '@material-ui/core/ListItemText'
import ListItemSecondaryAction from '@material-ui/core/ListItemSecondaryAction'
import Checkbox from '@material-ui/core/Checkbox'
import DeleteIcon from '@material-ui/icons/Delete'
import { withStyles } from '@material-ui/core'
import React from 'react'
import { Mutation } from 'react-apollo'
import { changeBookmarkReadFlag, removeBookmark, queryAllBookmarks } from './gql'

const styles = R.always({
    root: {
        color: 'inherit',
        textDecoration: 'none',
    },
    listItemContainer: {
        '&:hover $remove': {
            opacity: 1,
        },
    },
    linkRead: {
        color: '#00000080',
    },
    remove: {
        opacity: 0,
    },
})

function BookmarkItem({ classes, id, title, url, isRead }) {
    return (
        <Mutation
            mutation={changeBookmarkReadFlag}
            ignoreResults
            refetchQueries={[{ query: queryAllBookmarks }]}
        >
            {(updateBookmark) => {
                return (
                    <Mutation
                        mutation={removeBookmark}
                        ignoreResults
                        refetchQueries={[{ query: queryAllBookmarks }]}
                    >
                        {(removeBookmark) => {
                            return (
                                <ListItem
                                    component="a"
                                    href={url}
                                    className={classes.root}
                                    classes={{ container: classes.listItemContainer }}
                                >
                                    <Checkbox
                                        disableRipple
                                        checked={isRead}
                                        aria-label={isRead ? 'Mark unread' : 'Mark read'}
                                        tabIndex={-1}
                                        onChange={() => {
                                            updateBookmark({
                                                variables: {
                                                    id: id,
                                                    isRead: !isRead,
                                                },
                                                optimisticResponse: {
                                                    __typename: 'Mutation',
                                                    updateBookmark: {
                                                        id,
                                                        __typename: 'Bookmark',
                                                        isRead: !isRead,
                                                    },
                                                },
                                            })
                                        }}
                                    />
                                    <ListItemText
                                        primary={title}
                                        classes={{
                                            primary: classNames({
                                                [classes.linkRead]: isRead,
                                            }),
                                        }}
                                    />
                                    <ListItemSecondaryAction className={classes.remove}>
                                        <IconButton
                                            aria-label="remove item"
                                            onClick={() => {
                                                removeBookmark({
                                                    variables: {
                                                        id,
                                                    },
                                                })
                                            }}
                                        >
                                            <DeleteIcon />
                                        </IconButton>
                                    </ListItemSecondaryAction>
                                </ListItem>
                            )
                        }}
                    </Mutation>
                )
            }}
        </Mutation>
    )
}

BookmarkItem.propTypes = {
    classes: PropTypes.object.isRequired,
    id: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired,
    isRead: PropTypes.bool.isRequired,
}

export default withStyles(styles)(BookmarkItem)
