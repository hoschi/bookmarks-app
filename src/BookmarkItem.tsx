import * as React from 'react'
import { Bookmark } from './__generated__/graphql'
import {
    ListItem,
    Checkbox,
    ListItemText,
    ListItemSecondaryAction,
    IconButton,
} from '@mui/material'
import DeleteIcon from '@mui/icons-material/Delete'
import { useMutation } from '@apollo/client'
import {
    changeBookmarkReadFlag,
    removeBookmark as removeBookmarkMutation,
    queryAllBookmarks,
} from './gql'

export const BookmarkItem: React.FC<{ bookmark: Bookmark }> = ({ bookmark }) => {
    const { id, title, url, isRead } = bookmark
    const [updateBookmark, { error: updateBookmarkError }] = useMutation(changeBookmarkReadFlag, {
        refetchQueries: [queryAllBookmarks],
    })
    const [removeBookmark, { error: removeBookmarkError }] = useMutation(removeBookmarkMutation, {
        refetchQueries: [queryAllBookmarks],
    })

    if (updateBookmarkError) throw new Error(updateBookmarkError.toString())
    if (removeBookmarkError) throw new Error(removeBookmarkError.toString())

    return (
        <ListItem
            component="a"
            href={url}
            sx={{
                color: 'inherit',
                textDecoration: 'none',
                '&:hover + .MuiListItemSecondaryAction-root, & + .MuiListItemSecondaryAction-root:hover':
                    {
                        opacity: 1,
                    },
            }}
        >
            <Checkbox
                disableRipple
                checked={isRead}
                aria-label={isRead ? 'Mark unread' : 'Mark read'}
                tabIndex={-1}
                onChange={() => {
                    updateBookmark({
                        variables: {
                            id: id,
                            isRead: !isRead,
                        },
                        optimisticResponse: {
                            __typename: 'Mutation',
                            updateBookmark: {
                                id,
                                __typename: 'Bookmark',
                                isRead: !isRead,
                            },
                        },
                    })
                }}
            />
            <ListItemText
                primary={title}
                sx={{ color: isRead ? { color: '#00000080' } : undefined }}
            />
            <ListItemSecondaryAction
                sx={{
                    opacity: 0,
                }}
            >
                <IconButton
                    aria-label="remove item"
                    onClick={() => {
                        removeBookmark({
                            variables: {
                                id,
                            },
                        })
                    }}
                >
                    <DeleteIcon />
                </IconButton>
            </ListItemSecondaryAction>
        </ListItem>
    )
}
