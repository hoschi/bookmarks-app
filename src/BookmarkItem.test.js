import React from 'react'
import { shallow } from 'enzyme'
import IconButton from '@material-ui/core/IconButton'
import Checkbox from '@material-ui/core/Checkbox'
import { BookmarkItemStyled, adoptConfig } from './BookmarkItem'

test('render', function() {
    const updateBookmark = jest.fn()
    const removeBookmark = jest.fn()

    const actual = shallow(
        <BookmarkItemStyled
            id="123"
            title="my test title"
            url="https://medium.com/really-cool-article?campaign_id=45"
            isRead={false}
            updateBookmark={updateBookmark}
            removeBookmark={removeBookmark}
        />
    ).dive()
    expect(actual).toMatchSnapshot()

    expect(updateBookmark).toHaveBeenCalledTimes(0)
    expect(removeBookmark).toHaveBeenCalledTimes(0)

    actual.find(Checkbox).prop('onChange')()
    expect(updateBookmark).toHaveBeenCalledTimes(1)
    expect(removeBookmark).toHaveBeenCalledTimes(0)
    const updateParam = updateBookmark.mock.calls[0][0]
    expect(updateParam.variables).toEqual({
        id: '123',
        isRead: true,
    })
    expect(updateParam.optimisticResponse).toEqual({
        __typename: 'Mutation',
        updateBookmark: {
            __typename: 'Bookmark',
            id: '123',
            isRead: true,
        },
    })

    actual.find(IconButton).prop('onClick')()
    expect(updateBookmark).toHaveBeenCalledTimes(1)
    expect(removeBookmark).toHaveBeenCalledTimes(1)
    const removeParam = removeBookmark.mock.calls[0][0]
    expect(removeParam.variables).toEqual({
        id: '123',
    })
})

test('render with read item', function() {
    const updateBookmark = jest.fn()
    const removeBookmark = jest.fn()

    const actual = shallow(
        <BookmarkItemStyled
            id="123"
            title="my test title"
            url="https://medium.com/really-cool-article?campaign_id=45"
            isRead={true}
            updateBookmark={updateBookmark}
            removeBookmark={removeBookmark}
        />
    ).dive()
    expect(actual).toMatchSnapshot()
})

test('update Mutation', function() {
    const Component = adoptConfig.updateBookmark
    const actual = shallow(<Component render={jest.fn()} />)

    expect(actual).toMatchSnapshot()
})

test('remove Mutation', function() {
    const Component = adoptConfig.removeBookmark
    const actual = shallow(<Component render={jest.fn()} />)

    expect(actual).toMatchSnapshot()
})
