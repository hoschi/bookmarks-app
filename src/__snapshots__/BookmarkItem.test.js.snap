// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`remove Mutation 1`] = `
<Mutation
  ignoreResults={true}
  mutation={
    Object {
      "definitions": Array [
        Object {
          "directives": Array [],
          "kind": "OperationDefinition",
          "name": Object {
            "kind": "Name",
            "value": "RemoveBookmark",
          },
          "operation": "mutation",
          "selectionSet": Object {
            "kind": "SelectionSet",
            "selections": Array [
              Object {
                "arguments": Array [
                  Object {
                    "kind": "Argument",
                    "name": Object {
                      "kind": "Name",
                      "value": "id",
                    },
                    "value": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "id",
                      },
                    },
                  },
                ],
                "directives": Array [],
                "kind": "Field",
                "name": Object {
                  "kind": "Name",
                  "value": "removeBookmark",
                },
              },
            ],
          },
          "variableDefinitions": Array [
            Object {
              "kind": "VariableDefinition",
              "type": Object {
                "kind": "NonNullType",
                "type": Object {
                  "kind": "NamedType",
                  "name": Object {
                    "kind": "Name",
                    "value": "ID",
                  },
                },
              },
              "variable": Object {
                "kind": "Variable",
                "name": Object {
                  "kind": "Name",
                  "value": "id",
                },
              },
            },
          ],
        },
      ],
      "kind": "Document",
      "loc": Object {
        "end": 79,
        "source": Object {
          "body": "
    mutation RemoveBookmark($id: ID!) {
        removeBookmark(id: $id)
    }
",
          "locationOffset": Object {
            "column": 1,
            "line": 1,
          },
          "name": "GraphQL request",
        },
        "start": 0,
      },
    }
  }
  refetchQueries={
    Array [
      Object {
        "query": Object {
          "definitions": Array [
            Object {
              "directives": Array [],
              "kind": "OperationDefinition",
              "operation": "query",
              "selectionSet": Object {
                "kind": "SelectionSet",
                "selections": Array [
                  Object {
                    "arguments": Array [
                      Object {
                        "kind": "Argument",
                        "name": Object {
                          "kind": "Name",
                          "value": "sortField",
                        },
                        "value": Object {
                          "block": false,
                          "kind": "StringValue",
                          "value": "isRead",
                        },
                      },
                      Object {
                        "kind": "Argument",
                        "name": Object {
                          "kind": "Name",
                          "value": "sortOrder",
                        },
                        "value": Object {
                          "block": false,
                          "kind": "StringValue",
                          "value": "ASC",
                        },
                      },
                    ],
                    "directives": Array [],
                    "kind": "Field",
                    "name": Object {
                      "kind": "Name",
                      "value": "allBookmarks",
                    },
                    "selectionSet": Object {
                      "kind": "SelectionSet",
                      "selections": Array [
                        Object {
                          "arguments": Array [],
                          "directives": Array [],
                          "kind": "Field",
                          "name": Object {
                            "kind": "Name",
                            "value": "id",
                          },
                        },
                        Object {
                          "arguments": Array [],
                          "directives": Array [],
                          "kind": "Field",
                          "name": Object {
                            "kind": "Name",
                            "value": "title",
                          },
                        },
                        Object {
                          "arguments": Array [],
                          "directives": Array [],
                          "kind": "Field",
                          "name": Object {
                            "kind": "Name",
                            "value": "url",
                          },
                        },
                        Object {
                          "arguments": Array [],
                          "directives": Array [],
                          "kind": "Field",
                          "name": Object {
                            "kind": "Name",
                            "value": "isRead",
                          },
                        },
                      ],
                    },
                  },
                ],
              },
              "variableDefinitions": Array [],
            },
          ],
          "kind": "Document",
          "loc": Object {
            "end": 153,
            "source": Object {
              "body": "
    {
        allBookmarks(sortField: \\"isRead\\", sortOrder: \\"ASC\\") {
            id
            title
            url
            isRead
        }
    }
",
              "locationOffset": Object {
                "column": 1,
                "line": 1,
              },
              "name": "GraphQL request",
            },
            "start": 0,
          },
        },
      },
    ]
  }
>
  <Component />
</Mutation>
`;

exports[`render 1`] = `
<WithStyles(ListItem)
  className="BookmarkItem-root-1"
  classes={
    Object {
      "container": "BookmarkItem-listItemContainer-2",
    }
  }
  component="a"
  href="https://medium.com/really-cool-article?campaign_id=45"
>
  <WithStyles(Checkbox)
    aria-label="Mark read"
    checked={false}
    disableRipple={true}
    onChange={[Function]}
    tabIndex={-1}
  />
  <WithStyles(ListItemText)
    classes={
      Object {
        "primary": "",
      }
    }
    primary="my test title"
  />
  <WithStyles(ListItemSecondaryAction)
    className="BookmarkItem-remove-4"
  >
    <WithStyles(IconButton)
      aria-label="remove item"
      onClick={[Function]}
    >
      <pure(DeleteIcon) />
    </WithStyles(IconButton)>
  </WithStyles(ListItemSecondaryAction)>
</WithStyles(ListItem)>
`;

exports[`render with read item 1`] = `
<WithStyles(ListItem)
  className="BookmarkItem-root-1"
  classes={
    Object {
      "container": "BookmarkItem-listItemContainer-2",
    }
  }
  component="a"
  href="https://medium.com/really-cool-article?campaign_id=45"
>
  <WithStyles(Checkbox)
    aria-label="Mark unread"
    checked={true}
    disableRipple={true}
    onChange={[Function]}
    tabIndex={-1}
  />
  <WithStyles(ListItemText)
    classes={
      Object {
        "primary": "BookmarkItem-linkRead-3",
      }
    }
    primary="my test title"
  />
  <WithStyles(ListItemSecondaryAction)
    className="BookmarkItem-remove-4"
  >
    <WithStyles(IconButton)
      aria-label="remove item"
      onClick={[Function]}
    >
      <pure(DeleteIcon) />
    </WithStyles(IconButton)>
  </WithStyles(ListItemSecondaryAction)>
</WithStyles(ListItem)>
`;

exports[`update Mutation 1`] = `
<Mutation
  ignoreResults={true}
  mutation={
    Object {
      "definitions": Array [
        Object {
          "directives": Array [],
          "kind": "OperationDefinition",
          "name": Object {
            "kind": "Name",
            "value": "ChangeReadFlag",
          },
          "operation": "mutation",
          "selectionSet": Object {
            "kind": "SelectionSet",
            "selections": Array [
              Object {
                "arguments": Array [
                  Object {
                    "kind": "Argument",
                    "name": Object {
                      "kind": "Name",
                      "value": "id",
                    },
                    "value": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "id",
                      },
                    },
                  },
                  Object {
                    "kind": "Argument",
                    "name": Object {
                      "kind": "Name",
                      "value": "isRead",
                    },
                    "value": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "isRead",
                      },
                    },
                  },
                ],
                "directives": Array [],
                "kind": "Field",
                "name": Object {
                  "kind": "Name",
                  "value": "updateBookmark",
                },
                "selectionSet": Object {
                  "kind": "SelectionSet",
                  "selections": Array [
                    Object {
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "id",
                      },
                    },
                    Object {
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "isRead",
                      },
                    },
                  ],
                },
              },
            ],
          },
          "variableDefinitions": Array [
            Object {
              "kind": "VariableDefinition",
              "type": Object {
                "kind": "NonNullType",
                "type": Object {
                  "kind": "NamedType",
                  "name": Object {
                    "kind": "Name",
                    "value": "ID",
                  },
                },
              },
              "variable": Object {
                "kind": "Variable",
                "name": Object {
                  "kind": "Name",
                  "value": "id",
                },
              },
            },
            Object {
              "kind": "VariableDefinition",
              "type": Object {
                "kind": "NonNullType",
                "type": Object {
                  "kind": "NamedType",
                  "name": Object {
                    "kind": "Name",
                    "value": "Boolean",
                  },
                },
              },
              "variable": Object {
                "kind": "Variable",
                "name": Object {
                  "kind": "Name",
                  "value": "isRead",
                },
              },
            },
          ],
        },
      ],
      "kind": "Document",
      "loc": Object {
        "end": 161,
        "source": Object {
          "body": "
    mutation ChangeReadFlag($id: ID!, $isRead: Boolean!) {
        updateBookmark(id: $id, isRead: $isRead) {
            id
            isRead
        }
    }
",
          "locationOffset": Object {
            "column": 1,
            "line": 1,
          },
          "name": "GraphQL request",
        },
        "start": 0,
      },
    }
  }
  refetchQueries={
    Array [
      Object {
        "query": Object {
          "definitions": Array [
            Object {
              "directives": Array [],
              "kind": "OperationDefinition",
              "operation": "query",
              "selectionSet": Object {
                "kind": "SelectionSet",
                "selections": Array [
                  Object {
                    "arguments": Array [
                      Object {
                        "kind": "Argument",
                        "name": Object {
                          "kind": "Name",
                          "value": "sortField",
                        },
                        "value": Object {
                          "block": false,
                          "kind": "StringValue",
                          "value": "isRead",
                        },
                      },
                      Object {
                        "kind": "Argument",
                        "name": Object {
                          "kind": "Name",
                          "value": "sortOrder",
                        },
                        "value": Object {
                          "block": false,
                          "kind": "StringValue",
                          "value": "ASC",
                        },
                      },
                    ],
                    "directives": Array [],
                    "kind": "Field",
                    "name": Object {
                      "kind": "Name",
                      "value": "allBookmarks",
                    },
                    "selectionSet": Object {
                      "kind": "SelectionSet",
                      "selections": Array [
                        Object {
                          "arguments": Array [],
                          "directives": Array [],
                          "kind": "Field",
                          "name": Object {
                            "kind": "Name",
                            "value": "id",
                          },
                        },
                        Object {
                          "arguments": Array [],
                          "directives": Array [],
                          "kind": "Field",
                          "name": Object {
                            "kind": "Name",
                            "value": "title",
                          },
                        },
                        Object {
                          "arguments": Array [],
                          "directives": Array [],
                          "kind": "Field",
                          "name": Object {
                            "kind": "Name",
                            "value": "url",
                          },
                        },
                        Object {
                          "arguments": Array [],
                          "directives": Array [],
                          "kind": "Field",
                          "name": Object {
                            "kind": "Name",
                            "value": "isRead",
                          },
                        },
                      ],
                    },
                  },
                ],
              },
              "variableDefinitions": Array [],
            },
          ],
          "kind": "Document",
          "loc": Object {
            "end": 153,
            "source": Object {
              "body": "
    {
        allBookmarks(sortField: \\"isRead\\", sortOrder: \\"ASC\\") {
            id
            title
            url
            isRead
        }
    }
",
              "locationOffset": Object {
                "column": 1,
                "line": 1,
              },
              "name": "GraphQL request",
            },
            "start": 0,
          },
        },
      },
    ]
  }
>
  <Component />
</Mutation>
`;
